nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/AbstractGCPProcessor.java

```diff
@@ -0,0 +1,118 @@
+
+    public static final PropertyDescriptor PROJECT_ID = new PropertyDescriptor
+            .Builder().name("gcp-project-id")
+            .displayName("Project ID")
+            .description("Google Cloud Project ID")
+            .expressionLanguageSupported(true)
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor RETRY_COUNT = new PropertyDescriptor
+            .Builder().name("gcp-retry-count")
+            .displayName("Number of retries")
+            .description("How many retry attempts should be made before routing to the failure relationship.")
+            .defaultValue("6")
+            .expressionLanguageSupported(true)
+            .required(true)
+            .addValidator(StandardValidators.INTEGER_VALIDATOR)
+            .build();
+
+    /**
+     * Links to the {@link GCPCredentialsService} which provides credentials for this particular processor.
+     */
+    public static final PropertyDescriptor GCP_CREDENTIALS_PROVIDER_SERVICE = new PropertyDescriptor.Builder()
+            .name("gcp-credentials-provider-service")
+            .name("GCP Credentials Provider Service")
+            .description("The Controller Service used to obtain Google Cloud Platform credentials.")
+            .required(true)
+            .identifiesControllerService(GCPCredentialsService.class)
+            .build();
+
+
+    protected volatile CloudService cloudService;
+
+    protected CloudService getCloudService() {
+        return cloudService;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return ImmutableList.of(
+                GCP_CREDENTIALS_PROVIDER_SERVICE,
+                PROJECT_ID,
+                RETRY_COUNT
+        );
+    }
+
+
+    /**
+     * Retrieve credentials from the {@link GCPCredentialsService} attached to this processor.
+     * @param context the process context provided on scheduling the processor.
+     * @return GoogleCredentials for the processor to access.
+     * @see  <a href="https://developers.google.com/api-client-library/java/google-api-java-client/reference/1.20.0/com/google/api/client/googleapis/auth/oauth2/GoogleCredential">AuthCredentials</a>
+     */
+    private GoogleCredentials getGoogleCredentials(final ProcessContext context) {
+        final GCPCredentialsService gcpCredentialsService =
+                context.getProperty(GCP_CREDENTIALS_PROVIDER_SERVICE).asControllerService(GCPCredentialsService.class);
+        return gcpCredentialsService.getGoogleCredentials();
+    }
+
+    /**
+     * Assigns the cloud service client on scheduling.
+     * @param context the process context provided on scheduling the processor.
+     */
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
```

Comment:
It isn't clear to me just yet but as a thing to consider..  This 'cloudService' instance being set means this cloud service object must be thread safe should these processors have more than one thread and if this service is like a connection then it may be better to lazily init during onTrigger calls so that it can be recovered if the connection drops.  Open connections can be killed during onUnscheduled.  

What I'm suggesting here might not  matter if this object isn't what it seems in just scanning through the code.  I checked the google api for 'getService()' to see if I could better understand what it was and it wasnt revealing.

In any even this is a really substantive PR with a lot of well thought out areas and attention to details like LICENSE/NOTICE so big thanks to you.